README.md
# sqlite3-editor
Feel free to report any issues.


vscode/CHANGELOG.md
# Changelog
# v1.0.18
- Added context menu items.
- Fixed the behaviors of keyboard shortcuts around hidden columns and BLOB columns.
- Changed the BLOB editor to use file pickers to select filepaths instead of a textarea.
- Fixed a bug that caused the switches for the data types in the INSERT editor to be ignored in the resulting SQL query.

# v1.0.17
- Fixed a bug where the terminal opened by the Tools button would not reopen after being closed.

# v1.0.16
- Added an integration with sqlite-utils, which can be accessed via the Tools button.
- Fixed an issue where [SQLite's temporary files](https://www.sqlite.org/tempfiles.html) were not being deleted after the editor was closed, due to the connection to the database not being closed properly.
- Fixed an issue where auto-reloading was not functioning for databases that had WAL mode enabled.

# v1.0.15
- Reduced the minimum required version of SQLite bundled with Python from 3.37.0 to 3.8.7.1 by adding fallback code for `PRAGMA table_list`.

# v1.0.14
- Fixed #2: Added a check for minimum supported version of the SQLite bundled with Python.
- Fixed an issue where Python binaries installed from Microsoft Store is not recognized.

# v1.0.13
- Fixed an issue where certain characters were difficult to read in the dark theme.
- Added support for high-contrast-light themes.
- Fixed the code for the error handling of spawning Python.
- Fixed an issue where the animation triggered by clicks was not playing on slow machines due to a timing issue.

# v1.0.12
- Added a dark theme.

# v1.0.11
- Added the ability to hide columns.
- Added context menus to the table header and row numbers.
- Fixed an issue where the editor state was reset without prompting the user when using an editor other than UPDATE and clicking the right side of the table.
- Added an error handling code to display error messages when the Python binary specified in the settings.json does not exist.
- Added `ORDER BY rowid` to SELECT queries.

# v1.0.10
- Fixed a bug that caused the default table height to be set to 200 instead of 20.

# v1.0.9
- Fixed an issue of stuttering when changing the table's height.
- Fixed a bug where the input box in a cell would not respond to clicks.

# v1.0.8
- Added code to display runtime errors raised by Python.

# v1.0.7
- Fixed runtime errors in Python versions lower than 3.8.3.

# v1.0.6
- The extension now displays blank rows when there are few records, and switches to the INSERT editor when one of the cells is clicked.
- The "Commit changes?" dialog is now displayed for the INSERT, CREATE TABLE, CREATE INDEX, and custom query editor, as well as the UPDATE editor.
- Tooltips on icon buttons are now displayed immediately when hovered to improve accessibility.

# v1.0.5
- Fixed a bug where columns could not be resized while updating a cell.
- Added a cancel button to the "Commit changes?" dialog and made it focus the "Commit" button when opened.
- Fixed the behaviors of the Enter and Tab keys while updating a cell.
- Added LICENSE (MIT) and `vendor.LICENSE.txt`.

# v1.0.4
- Fixed a problem where undo/redo were not working on inputs.

# v1.0.3
- Fixed #1: *.sqlite and *.sqlite3 are now recognized as SQLite databases.

# v1.0.2
- Fixed a bug introduced in v1.0.1.

# v1.0.1
- Added keyboard shortcuts to move selection.


ui/util.ts
import create, { StoreApi, UseBoundStore } from "zustand"

export const BigintMath = {
    max: (...args: bigint[]) => args.reduce((prev, curr) => curr > prev ? curr : prev),
    min: (...args: bigint[]) => args.reduce((prev, curr) => curr < prev ? curr : prev),
}

/**
 * Attempts to find the first element that matches the given CSS selector using `document.querySelector`. If the element is not found, it sets up a MutationObserver to observe the body element and its descendant nodes for mutations. If a matching element is added as a result of a mutation, the MutationObserver is disconnected and the element is returned. The MutationObserver waits for at most one second for a matching element to be added.
 * @param selector - The CSS selector to use.
 * @returns The first element that matches the given selector, or `null` if no such element is found after one second of observing for mutations.
 */
export const querySelectorWithRetry = <T extends Element = Element>(selector: string) => new Promise<T | null>((resolve) => {
    // Attempt to find the element using document.querySelector
    const result = document.querySelector<T>(selector)
    if (result) {
        // If the element was found, return it
        resolve(result)
        return
    }

    // Create a MutationObserver to listen for mutations to the body element or its descendant nodes
    const observer = new MutationObserver((mutations) => {
        // For each mutation
        for (const mutation of mutations) {
            for (const child of mutation.addedNodes) {
                if (!(child instanceof Element)) { continue }
                const result = child.querySelector<T>(selector)
                if (result) {
                    // Disconnect the observer and return the element
                    observer.disconnect()
                    clearTimeout(timeout)
                    resolve(result)
                    return
                }
            }
        }
    })

    // If the element was not found, start observing for mutations
    observer.observe(document.body, { childList: true, subtree: true })

    // Set a timeout to stop observing after one second
    const timeout = setTimeout(() => {
        observer.disconnect()
        resolve(null)
    }, 1000)
})

type Store<T> = UseBoundStore<StoreApi<T>>
export const createStore = <S, A>(name: string, initialState: S, actions: (set: Store<S>["setState"], get: Store<S>["getState"]) => A) =>
    hmr(name, create<S & A>()((set, get) => ({ ...initialState, ...actions(set, get) })))

const hmr = <T>(key: string, store: T): T =>
    // @ts-ignore
    import.meta.env.DEV ? (window as any)[key] ?? ((window as any)[key] = store) : store


ui/remote.ts
import { Packr, Unpackr } from "msgpackr"
import { escapeSQLIdentifier, useTableStore } from "./table"

export type TableInfo = { cid: bigint, dflt_value: bigint | string | null, name: string, notnull: bigint, type: string, pk: bigint }[]
export type UniqueConstraints = { primary: boolean, columns: string[] }[]
export type SQLite3Value = string | number | bigint | Uint8Array | null
export type TableListItem = { name: string, type: "table" | "view" | "shadow" | "virtual", wr: bigint, strict: bigint }
export type ForeignKeyList = { id: bigint, seq: bigint, table: string, from: string, to: string, on_update: string, on_delete: string, match: string }[]

export type Message = { data: { /* preact debugger also uses message events */ type: "sqlite3-editor-server" } & ({ requestId: undefined } | { requestId: number } & ({ err: string } | { body: Uint8Array })) }

const packr = new Packr({ useRecords: false, preserveNumericTypes: true })
const unpackr = new Unpackr({ largeBigIntToFloat: false, int64AsNumber: false, mapsAsObjects: true, useRecords: true, preserveNumericTypes: true })

const vscode = window.acquireVsCodeApi?.()

const queue: { start: number }[] = []

const loop = () => {
    if (queue.length > 0 && Date.now() - queue[0]!.start > 1000) {
        document.body.classList.add("querying")
    } else {
        document.body.classList.remove("querying")
    }
    requestAnimationFrame(loop)
}
loop()

type PostOptions = { withoutLogging?: boolean }

/** Send the data to the server with fetch(), or to the extension host with vscode.postMessage(). */
export const post = async (url: string, body: unknown, opts: PostOptions = {}) => {
    const id = { start: Date.now() }
    queue.push(id)
    if (vscode !== undefined) {
        return new Promise<unknown>((resolve, reject) => {
            const requestId = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER)
            vscode.postMessage({ requestId, path: url, body: packr.pack(body) })
            const callback = ({ data }: Message) => {
                if (data.type !== "sqlite3-editor-server" || data.requestId !== requestId) { return }
                window.removeEventListener("message", callback)
                if ("err" in data) {
                    if (!opts.withoutLogging) { useTableStore.getState().addErrorMessage(data.err) }
                    reject(new Error(data.err))
                    return
                }
                resolve(unpackr.unpack(data.body))
            }
            window.addEventListener("message", callback)
        }).finally(() => {
            queue.splice(queue.indexOf(id)!, 1)
        })
    } else {
        try {
            let res: Response
            try {
                res = await fetch("http://localhost:8080" + url, { method: "POST", body: packr.pack(body), headers: { "Content-Type": "application/octet-stream" } })
            } catch (err) {
                if (!opts.withoutLogging) { useTableStore.getState().addErrorMessage(typeof err === "object" && err !== null && "message" in err ? err.message + "" : err + "") }
                throw err
            }
            if (!res.ok) {
                const message = await res.text()
                if (!opts.withoutLogging) { useTableStore.getState().addErrorMessage(message) }
                throw new Error(message)
            }
            return unpackr.unpack(new Uint8Array(await res.arrayBuffer()))
        } finally {
            queue.splice(queue.indexOf(id)!, 1)
        }
    }
}

/** A local copy of the persisted data in the server. */
let state: Record<string, unknown> = {}

/** Retrieves and stores the persisted data in the server. */
export const downloadState = (opts: PostOptions = {}): Promise<void> => post("/downloadState", {}, opts).then((value) => { state = value })

/** Stores the value with the given key to the server. */
export const setState = async <T>(key: string, value: T, opts: PostOptions = {} = {}) => {
    state[key] = value
    await post("/setState", { key, value }, opts)
}

/** Returns the value associated with the given key from local copy of the persisted data in the server. `downloadState()` have be called beforehand. */
export const getState = <T>(key: string): T | undefined => {
    return state[key] as T | undefined
}

type QueryResult<T extends string> = Promise<{
    columns: string[]
    records:
    T extends `SELECT ${string}` ? Record<string, SQLite3Value>[] :
    T extends `PRAGMA pragma_list` ? { name: string[] } :
    (Record<string, SQLite3Value>[] | undefined)
}>

/** Queries the database, and commits if `mode` is "w+". */
export const query = <T extends string>(query: T, params: readonly SQLite3Value[], mode: "r" | "w+", opts: PostOptions = {}): QueryResult<T> =>
    post(`/query`, { query, params, mode }, opts)

/** Opens an integrated terminal and types the text. */
export const openTerminal = async (text: string, opts: PostOptions = {}) => {
    await post("/openTerminal", { text }, opts)
}

/** https://stackoverflow.com/a/1604121/10710682 */
export const existsTable = async (tableName: string, opts: PostOptions = {}) =>
    (await query(`SELECT name FROM sqlite_master WHERE type='table' AND name=?`, [tableName], "r", opts)).records.length > 0

/** https://stackoverflow.com/questions/20979239/how-to-tell-if-a-sqlite-column-is-autoincrement */
export const hasTableAutoincrementColumn = async (tableName: string, opts: PostOptions = {}) =>
    await existsTable("sqlite_sequence", opts) &&  // sqlite_sequence doesn't exist when the database is empty.
    !!((await query(`SELECT COUNT(*) AS count FROM sqlite_sequence WHERE name = ?`, [tableName], "r", opts)).records[0]?.count)

export type IndexInfo = { seqno: bigint, cid: bigint, name: string }[]

/** Queries `PRAGMA index_info(indexName)`. */
export const getIndexInfo = async (indexName: string, opts: PostOptions = {}) =>
    (await query(`PRAGMA index_info(${escapeSQLIdentifier(indexName)})`, [], "r", opts)).records as IndexInfo

export type IndexList = { seq: bigint, name: string, unique: 0n | 1n, origin: "c" | "u" | "pk", partial: 0n | 1n }[]

/** Queries `PRAGMA index_list(tableName)`. */
export const getIndexList = async (tableName: string, opts: PostOptions = {}) =>
    (await query(`PRAGMA index_list(${escapeSQLIdentifier(tableName)})`, [], "r", opts)).records as IndexList

/** Queries `PRAGMA table_info(tableName)`. */
export const getTableInfo = async (tableName: string, opts: PostOptions = {}) =>
    (await query(`PRAGMA table_info(${escapeSQLIdentifier(tableName)})`, [], "r", opts)).records as TableInfo


export const getForeignKeyList = async (tableName: string, opts: PostOptions = {}) =>
    (await query(`PRAGMA foreign_key_list(${escapeSQLIdentifier(tableName)})`, [], "r", opts)).records as ForeignKeyList

export let isSQLiteOlderThan3_37 = false

/** Lists tables in the database, excluding internal tables. */
export const getTableList = async (opts: PostOptions = {}) => {
    const tableList = await query("PRAGMA table_list", [], "r", opts)  // Requires SQLite >= 3.37.0
    if (!tableList) {  // if using SQLite < 3.37.0
        isSQLiteOlderThan3_37 = true
        const tableList: TableListItem[] = []
        for (const { name, type, sql } of (await query(`SELECT name, type, sql FROM sqlite_master WHERE NOT (name LIKE "sqlite\\_%" ESCAPE "\\") AND type IN ("table", "view", "shadow", "virtual")`, [], "r", opts)).records as { name: string, type: "table" | "view" | "shadow" | "virtual", sql: string }[]) {
            tableList.push({
                name,
                type,
                wr: /WITHOUT\s*ROWID[^()]*$/.test(sql) ? 1n : 0n,
                strict: 0n, // The STRICT option is introduced in SQLite 3.37.
            })
        }
        return tableList
    }
    return (tableList.records as TableListItem[])
        .filter(({ name }) => !name.startsWith("sqlite_"))  // https://www.sqlite.org/fileformat2.html#intschema
}

/** Retrieves the table schema. */
export const getTableSchema = async (tableName: string, opts: PostOptions = {}) =>
    // sqlite_master is renamed to sqlite_master in 3.33.0, but sqlite_master is kept as an alias https://sqlite.org/forum/forumpost/889b616b87a7020a
    (await query(`SELECT sql FROM sqlite_master WHERE name = ?`, [tableName], "r", opts)).records[0]?.sql as string | undefined

/** Retrieves the index schema. */
export const getIndexSchema = async (indexName: string, opts: PostOptions = {}) =>
    (await query(`SELECT sql FROM sqlite_master WHERE type = ? AND name = ?`, ["index", indexName], "r", opts)).records[0]?.sql as string | null | undefined


vscode/server.py
import argparse
import re
import sqlite3
import sys
import urllib.parse
from typing import Any

from umsgpack import pack, unpack


def find_widget_regexp(text: str, pattern: str, whole_word: int, case_sensitive: int):
    try:
        return 0 if re.search(f"\\b(?:{pattern})\\b" if whole_word else pattern, text, 0 if case_sensitive else re.RegexFlag.I) is None else 1
    except re.error:  # Invalid regular expressions
        return 0

class Server:
    def __init__(self, database_filepath: str):
        self.readonly_connection = sqlite3.connect("file:" + urllib.parse.quote(database_filepath) + "?mode=ro", uri=True)
        self.readwrite_connection = sqlite3.connect(database_filepath)

        if sys.version_info >= (3, 8, 3):  # `deterministic` is added in Python 3.8.3  https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.create_function
            self.readonly_connection.create_function("find_widget_regexp", 4, find_widget_regexp, deterministic=True)
            self.readwrite_connection.create_function("find_widget_regexp", 4, find_widget_regexp, deterministic=True)
        else:
            self.readonly_connection.create_function("find_widget_regexp", 4, find_widget_regexp)
            self.readwrite_connection.create_function("find_widget_regexp", 4, find_widget_regexp)

    def handle(self, request_body_filepath: str, response_body_filepath: str):
        def return_(status: int, value: Any):
            if status == 200:
                with open(response_body_filepath, "wb") as f:
                    pack(value, f)
            else:
                with open(response_body_filepath, "w") as f:
                    f.write(value)
            return status

        try:
            # request_body: { query: string, params: (null | number | bigint | string | Uint8Array | Buffer)[], mode: "w+" | "r" }
            with open(request_body_filepath, "rb") as f:
                request_body = unpack(f)

            try:
                if request_body["mode"] == "w+":
                    # read-write
                    with self.readwrite_connection as con:
                        con.execute(request_body["query"], request_body["params"])
                        return return_(200, None)
                else:
                    # read-only
                    with self.readonly_connection as con:
                        cursor = con.execute(request_body["query"], request_body["params"])
                    if cursor.description is not None:
                        columns = [desc[0] for desc in cursor.description]
                        return return_(200, {"columns": columns, "records": [{k: v for k, v in zip(columns, record)} for record in cursor.fetchall()]})
                    else:
                        return return_(200, None)
            except Exception as err:
                return return_(400, f"{err}\nQuery: {request_body['query']}\nParams: {request_body['params']}")
        except Exception as err:
            return return_(400, str(err))

    def close(self):
        self.readonly_connection.close()

        # Create a noop checkpoint to delete WAL files. https://www.sqlite.org/wal.html#avoiding_excessively_large_wal_files
        with self.readwrite_connection as con:
            con.execute("SELECT * FROM sqlite_master LIMIT 1").fetchall()
        self.readwrite_connection.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--database-filepath", type=str, required=True)
    parser.add_argument("--request-body-filepath", type=str, required=True)
    parser.add_argument("--response-body-filepath", type=str, required=True)
    args = parser.parse_args()
    server = Server(args.database_filepath)
    try:
        while True:
            command = input().strip()
            if command == "handle":
                print(server.handle(args.request_body_filepath, args.response_body_filepath), flush=True)
            elif command == "close":
                break
    finally:
        server.close()


ui/vite.config.js
import { defineConfig } from "vite"
import preact from "@preact/preset-vite"
import license from "rollup-plugin-license"
import path from "path"

export default defineConfig({
    plugins: [
        preact(),

        // https://github.com/vitejs/vite/discussions/7722#discussioncomment-4007436
        license({
            thirdParty: {
                output: path.resolve(__dirname, "./dist/assets/vendor.LICENSE.txt"),
            },
        }),
    ],
    build: {
        target: "esnext",
    },
    esbuild: {
        logOverride: { 'this-is-undefined-in-esm': 'silent' },

        // https://github.com/vitejs/vite/discussions/7722#discussioncomment-4007436
        banner: '/*! licenses: /assets/vendor.LICENSE.txt */',
        legalComments: 'none',
    },
    resolve: {
        alias: {
            "node:buffer": 'buffer/',  // file-type imports "node:buffer".
        }
    },
})


vscode/extension.ts
import vscode from "vscode"
import { spawn, spawnSync } from "child_process"
import { temporaryFile } from "tempy"
import fs from "fs"
import path from "path"
import which from "which"
import { Packr, Unpackr } from "msgpackr"
import os from "os"

const packr = new Packr({ useRecords: false, preserveNumericTypes: true })
const unpackr = new Unpackr({ largeBigIntToFloat: false, int64AsNumber: false, mapsAsObjects: true, useRecords: true, preserveNumericTypes: true })

class LocalPythonClient {
    readonly #p
    readonly #requestBody = temporaryFile({ extension: "msgpack" })
    readonly #responseBody = temporaryFile({ extension: "msgpack" })
    #closed: false | "success" | "error" = false

    #resolve!: (data: Buffer) => void
    #reject!: (message: Error) => void

    constructor(pythonPath: string, serverScriptPath: string, databasePath: string) {
        this.#p = spawn(pythonPath, [
            serverScriptPath,
            "--database-filepath", databasePath,
            "--request-body-filepath", this.#requestBody,
            "--response-body-filepath", this.#responseBody,
        ])
        this.#p.on("error", (err) => {
            vscode.window.showErrorMessage(err.message)
        })
        let errors = ""
        this.#p.stderr.on("data", (err: Buffer) => {
            const errStr = err.toString()
            if (errStr.includes("Traceback (") || this.#closed === "error") {
                // Show the error if it is a runtime error.
                vscode.window.showErrorMessage(errStr)
            } else {
                console.error(errStr)
                errors += errStr + "\n"
            }
        })
        this.#p.stdout.on("data", (data: Buffer) => {
            const status = +data.toString().trim()
            if (status === 400) {
                this.#reject(new Error(fs.readFileSync(this.#responseBody).toString()))
            } else if (status === 200) {
                this.#resolve(fs.readFileSync(this.#responseBody))
            }
        })
        this.#p.on("exit", (code) => {
            if (this.#closed) { return }
            this.#closed = "error"
            vscode.window.showErrorMessage(`Process exited: ${code}\n${errors}`)
        })
    }

    #queue = new Array<() => void>()
    request(body: Buffer | Uint8Array, resolve: (data: Buffer) => void, reject: (err: Error) => void) {
        const job = () => {
            fs.writeFileSync(this.#requestBody, body)
            this.#resolve = (data) => { resolve(data); this.#queue.shift(); this.#queue[0]?.() }
            this.#reject = (err) => { reject(err); this.#queue.shift(); this.#queue[0]?.() }
            this.#p.stdin.write("handle\n")
        }
        this.#queue.push(job)
        if (this.#queue.length === 1) {
            this.#queue[0]!()
        }
    }

    close() {
        this.#closed ||= "success"
        this.#p.stdin.write("close\n")
        this.#p.once("close", () => {
            fs.rmSync(this.#requestBody, { force: true })
            fs.rmSync(this.#responseBody, { force: true })
        })
    }
}

/**
 * Calls spawnSync() returns the child process's stdout as a string.
 * If the child process returns a non-zero exit code, or if an error occurs, the `or` value is returned.
 */
const spawnSyncOr = (command: string, args: readonly string[], or: string) => {
    try {
        const p = spawnSync(command, args)
        if (p.status !== 0) { return or }
        return p.stdout.toString()
    } catch (err) {
        return or
    }
}

const supportedPythonVersion = [3, 6] as const
const supportedSQLiteVersion = [3, 8]

const checkPythonVersion = (filepath: string) => {
    const pythonVersionCheck = spawnSyncOr(filepath, ["-c", `import sys; print(sys.version_info >= (${supportedPythonVersion[0]}, ${supportedPythonVersion[1]}))`], "")
    if (!pythonVersionCheck.includes("True")) { return false }
    const sqliteVersionCheck = spawnSyncOr(filepath, ["-c", `import sqlite3; print(tuple(map(int, sqlite3.sqlite_version.split("."))) >= (${supportedSQLiteVersion[0]}, ${supportedSQLiteVersion[1]}))`], "")
    if (!sqliteVersionCheck.includes("True")) { return false }
    return true
}

const findPython = async () => {
    const [major, minor] = supportedPythonVersion

    // Microsoft store
    for (const name of [
        ...[...Array(10).keys()].map((x) => `python${major}.${x + minor}`).reverse(),
        `python${major}`,
    ]) {
        // fs.existsSync(filepath) throws permission errors so we need to try executing the binary.
        const filepath = `${os.homedir()}\\AppData\\Local\\Microsoft\\WindowsApps\\${name}.exe`
        if (!spawnSyncOr(filepath, ["--version"], "").includes("Python")) { continue }
        if (!checkPythonVersion(filepath)) { continue }
        return filepath
    }

    // Other installations
    for (const name of [
        ...[...Array(10).keys()].map((x) => `python${major}.${x + minor}`).reverse(),
        `python${major}`,
        "python",
        "py",
    ]) {
        try {
            const filepath = await which(name)  // which() doesn't find applications in the WindowsApps directory.
            if (!checkPythonVersion(filepath)) { continue }
            return filepath
        } catch (err) {
            if ((err as any).code !== "ENOENT") {
                console.error(err)
            }
        }
    }
    return null
}

export const activate = (context: vscode.ExtensionContext) => {
    let terminal: vscode.Terminal | undefined
    context.subscriptions.push(
        {
            dispose: () => {
                terminal?.dispose()
                terminal = undefined
            }
        },
        vscode.window.registerCustomEditorProvider("sqlite3-editor.editor", {
            async openCustomDocument(uri, openContext, token) {
                const pythonPath = (vscode.workspace.getConfiguration("sqlite3-editor").get<string>("pythonPath") || await findPython())
                if (!pythonPath) {
                    const msg = `Could not find a Python >=${supportedPythonVersion[0]}.${supportedPythonVersion[1]} binary compiled with SQLite >=${supportedSQLiteVersion[0]}.${supportedSQLiteVersion[1]} .Install one from https://www.python.org/ or your OS's package manager (Microsoft Store, brew, apt, etc.).`
                    vscode.window.showErrorMessage(msg)
                    throw new Error(msg)
                }
                if (uri.scheme === "file") {
                    const conn = new LocalPythonClient(pythonPath, context.asAbsolutePath("server.py"), uri.fsPath)
                    // Watch the database file
                    const watcher = fs.watch(uri.fsPath)

                    // Watch the -wal file. We use setInterval instead of fs.watch because fs.watch doesn't accept non-existent files.
                    let mtime = 0
                    const timer = setInterval(() => {
                        const walFile = `${uri.fsPath}-wal`
                        try {
                            const newMtime = fs.statSync(walFile).mtimeMs
                            if (mtime !== newMtime) {
                                watcher.emit("change", walFile)
                            }
                            mtime = newMtime
                        } catch (err) {
                            if (!(err instanceof Error && "code" in err && err.code === "ENOENT")) {
                                console.error(err) // Show the error if it isn't an ENOENT
                            }
                        }
                    }, 1000)

                    return {
                        uri,
                        unsupportedScheme: false,
                        dispose: () => {
                            conn.close()
                            watcher.removeAllListeners()
                            watcher.close()
                            clearInterval(timer)
                        },
                        conn,
                        watcher,
                        pythonPath,
                    }
                } else {  // unsupportedScheme such as the diff view.
                    return {
                        uri,
                        unsupportedScheme: true,
                        dispose: () => { },
                    }
                }
            },
            async resolveCustomEditor(document, webviewPanel, token) {
                if (document.unsupportedScheme) {
                    if (document.uri.scheme === "git") {
                        let options = ""
                        try {
                            if (JSON.parse(decodeURIComponent(document.uri.query)).ref === "") {
                                options = " --staged"
                            }
                        } catch (err) { console.error(err) }
                        webviewPanel.webview.html = `The sqlite3 - editor extension doesn't support the git-diff view. Use the following command instead.<br /><code>${escapeHTML(`git -c 'diff.default.textconv = echo.dump | sqlite3' diff${options} '${document.uri.fsPath}'`)}</code>`
                    } else {
                        webviewPanel.webview.html = `Unsupported file scheme: ${escapeHTML(document.uri.scheme)}`
                    }
                    return
                }
                webviewPanel.webview.options = {
                    enableScripts: true,
                    localResourceRoots: [vscode.Uri.joinPath(context.extensionUri, "webview")],
                }
                webviewPanel.webview.html = (await vscode.workspace.fs.readFile(vscode.Uri.joinPath(context.extensionUri, "webview", "index.html"))).toString()
                    .replace(/((?:src|href)=")\/([^"]+)(")/g, (_, m1, m2, m3) => m1 + webviewPanel.webview.asWebviewUri(vscode.Uri.joinPath(context.extensionUri, "webview", m2)).toString() + m3)

                context.subscriptions.push(webviewPanel.webview.onDidReceiveMessage(({ requestId, path: url, body }: { requestId: number, path: string, body: Uint8Array }) => {
                    const res = {
                        send: (buf: Buffer | Uint8Array) => { webviewPanel.webview.postMessage({ type: "sqlite3-editor-server", requestId, body: new Uint8Array(buf) }) },
                        status: (code: 400) => ({ send: (text: string) => { webviewPanel.webview.postMessage({ type: "sqlite3-editor-server", requestId, err: text }) } })
                    }
                    switch (url) {
                        case "/setState": {
                            const { key, value } = packr.unpack(body) as { key: string, value: unknown }
                            context.workspaceState.update(key, value)
                                .then(() => { res.send(packr.pack(null)) }, (err) => { res.status(400).send((err as Error).message) })
                            break
                        } case "/downloadState": {
                            res.send(packr.pack(Object.fromEntries(context.workspaceState.keys().map((key) => [key, context.workspaceState.get(key)]))))
                            break
                        } case "/openTerminal": {
                            const { text } = packr.unpack(body) as { text: string }
                            if (!terminal || terminal.exitStatus !== undefined) {
                                terminal = vscode.window.createTerminal("SQLite3 Editor")
                            }
                            const pipList = JSON.parse(spawnSyncOr(document.pythonPath, ["-m", "pip", "list", "--format=json"], "[]")) as { name: string, version: string }[]
                            terminal.sendText(text
                                .replaceAll("{{install sqlite-utils &&}}", pipList.some(({ name, version }) => name === "sqlite-utils" && +version.split(".")[0]! >= 3) ? "" : "{{pythonPath}} -m pip install -qU --user sqlite-utils && ")
                                .replaceAll("{{pythonPath}}", escapeShell(document.pythonPath))
                                .replaceAll("{{databasePath}}", escapeShell(document.uri.fsPath)), false)
                            terminal.show()
                            res.send(packr.pack(null))
                            break
                        } default:
                            document.conn.request(body, (body) => res.send(body), (err) => { res.status(400).send((err as Error).message) })
                            break
                    }
                }))

                document.watcher.on("change", () => {
                    webviewPanel.webview.postMessage({ type: "sqlite3-editor-server" })
                })
            },
        } /* satisfies */ as vscode.CustomReadonlyEditorProvider<vscode.Disposable & (
            | {
                uri: vscode.Uri
                unsupportedScheme: false
                conn: LocalPythonClient
                watcher: fs.FSWatcher
                pythonPath: string
            }
            | {
                uri: vscode.Uri
                unsupportedScheme: true
            }
        )>, {
            supportsMultipleEditorsPerDocument: true,
            webviewOptions: {
                enableFindWidget: false,
                retainContextWhenHidden: true,
            },
        }),
    )
}

export const deactivate = () => { }

const escapeHTML = (t: string) => t
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#039;")

const escapeShell = (s: string) => `'${s.replaceAll("'", "'\\''")}'`
